server/server.js

```diff
@@ -56,10 +63,34 @@ Server.prototype.terminate = function(done) {
   });
 
   this.sentry.stop();
-  this.socketServer.close();
+
+  if (this.socketServer) {
+    this.socketServer.close();  
+  }
+  
   Core.Persistence.disconnect(done);
 };
 
+Server.prototype.use = function (middleware) {
+  this._middleware.push(middleware);
+};
+
+Server.prototype._runMiddleware = function() {
+  var context = arguments[0],
+      args = [].slice.call(arguments, 1, -1),
+      callback = [].slice.call(arguments, -1)[0];
+
+  var process = function (middleware, next) {
+    if (middleware[context]) {
+      middleware[context].apply(middleware, args.concat(next));
+    } else {
+      next();
+    }
+  };
+
+  async.each(this._middleware, process, callback);
+};
```

Comment:
This will get smaller once we extract all the authorization and limiting code. Bear with me.